main/src/ca/uwaterloo/flix/language/phase/ClosureConv.scala

```diff
@@ -101,6 +105,26 @@ object ClosureConv {
     case Expr.LetRec(sym, e1, e2, tpe, purity, loc) =>
       Expr.LetRec(sym, visitExp(e1), visitExp(e2), tpe, purity, loc)
 
+    case Expr.LocalDef(sym, fparams, exp1, exp2, tpe, purity, loc) =>
+      // Step 1: Compute the free variables in the body expression.
+      //         (Remove the variables bound by the function itself).
+      val bound = sym :: fparams.map(_.sym)
+      val fvs = filterBoundVars(freeVars(exp1), bound).toList
+
+      // Step 2: Convert the free variables into a new parameter list and substitution.
+      val (cloParams, subst) = getFormalParamsAndSubst(fvs, loc)
+
+      // Step 3: Rewrite every recursive call to include free vars and then replace
+      //         every occurrence of the free vars with their substitution.
+      val e1 = visitExp(applySubst(rewriteApplyLocalDef(exp1, sym, fvs), subst))
```

Comment:
Does it matter if you `rewriteApplyLocalDef` first or `applySubst` first? it feels like both would work. Was just curious

Reply:
The order matters since we rewrite with `fvs`, i.e., introducing them as args, and then renaming them via the substitution afterwards. My idea was that I wanted to first do one rewriting under the same variable naming scheme and then afterwards apply the substitution. I guess you could rewrite with the variables in `cloParams` directly but I like the consistency with this and we have to apply the substitution anyway so we do not really gain anything by being "smart" that way.

Reply:
`def f(x) = f(x +  y)` becomes `def f(q, x) = f(q, x + q)` right? 

So subst first would be `f(x + y) --> f(x + q) --> f(q, x + q)`

and rewrite first would be `f(x + y) --> f(y, x + y) --> f(q, x + q)`

I think subst first feels better to me, but I don't have any concrete reason

Reply:
I see. Yeah and getFormalParamsAndSubst doesnt return the new names. good point

Reply:
but yeah an example in docs somewhere in this file feels required